In [6]:
import pandas as pd

In [154]:
pip install pyspellchecker

Note: you may need to restart the kernel to use updated packages.


In [7]:
data = pd.read_excel("C:/Users/jw156/Downloads/Database Engineer Question 2 JDW Data.xlsx")

In [155]:
from spellchecker import SpellChecker

# There is a spelling mistake in the title of products in the dresses subcategory

In [189]:
spell_checker('the catz sat on the mat')

'spelling_mistake'

In [192]:
def spell_checker(words):
    new_words = words.split(' ')
    if len(new_words) < 0:
        return 'empty'
    spell = SpellChecker()
    misspelled = spell.unknown(new_words)
    if len(misspelled) >= 1:
        return ('spelling_mistake')

In [ ]:
data['spelling'] = data['title_g'].apply(spell_checker)

# There are products that are in the wrong category in the swimwear subcategory

In [147]:
swimwear = data[data['categories-href']== 'https://www.jdwilliams.co.uk/shop/c/womens/swimwear?cm_sp=JDW-Womens-_-ShopByCategory-_-SwimwearandBeachwear']

In [142]:
x = data[data['category_3']== 'Swimwear & Beachwear']

In [143]:
x.categories.unique()

array([], dtype=object)

In [50]:
swimwear = data[data['categories'] =='Swimwear & Beachwear']

In [57]:
swimwear['categories-href'].unique()

array(['https://www.jdwilliams.co.uk/shop/c/womens/swimwear?cm_sp=JDW-Womens-_-ShopByCategory-_-SwimwearandBeachwear'],
      dtype=object)

In [ ]:
# swimwear.category_3[1]

cannot find any swimwear that is within the wrong category

# Products with missing images from all subcategories

In [120]:
def missing_image(href):
    if pd.isnull(href) == True:
        return 'empty'

In [121]:
data['miss_image'] = data['Image-src'].apply(missing_image)

In [122]:
data['miss_image'].nunique()

1

# Products with missing descrip"ons from all subcategories

In [116]:
data['category_2'].isnull().sum()

22147

In [117]:
def missing_image(href, lref, zref, yref):
    if pd.isnull(href) == True or pd.isnull(lref) == True or pd.isnull(zref) == True or pd.isnull(yref) == True:
        return 'empty'

In [118]:
data['miss-descr'] = data.apply(lambda x: missing_image(x.category_0,x.category_1, x.category_2, x.category_3), axis=1)

In [119]:
data['miss-descr'].unique()

array(['empty', None], dtype=object)

#  Products where the stars are less than 2 and the reviews are greater than 3vi

In [184]:
def starts_reviews(star, reviews):
    if star < 2 and reviews > 3:
        return 'matched'

In [185]:
data['stars and reviews'] = data.apply(lambda x: starts_reviews(x.reviews_p,x.stars_p), axis=1)

In [186]:
data['stars and reviews'].unique()

array([None, 'matched'], dtype=object)

In [ ]:
data['web-scraper-start-url'][1]

# Give 3 examples of products that have prices you would consider to be “obscure”

In [ ]:
calculate the now

In [128]:
data['price_g'].unique()

array([22, nan, 24.5, 25, 130, 16, 11, 24, 15.25, 14, 26, 28, 6, 9, 35,
       15.2, 20, 18, 13, 32, 27, 10, 45.5, 30, 37, 41.5, 29, 45, 50, 17.5,
       15, 27.5, 14.3, 34, 17.6, 38, 44, 62, 36, 46, 42, 12, 19, 79, 39,
       21, 84, 40, 78, 259.99, 14.4, 86, 54, 8, 55, 57, 17, 49, 33, 48,
       10.4, 8.8, 9.6, 23, 49.99, 160, 60, 56, 149, 110, 59, 150, 99, 140,
       63, 65, 89, 70, 75, 95, 120, 269, 43, 80, 85, 69, 58, 170, 199,
       52.5, 105, 145, 100, 90, 250, 165, 125, 79.2, 200, 53, 135, 115, 7,
       21.99, 19.99, 'From £8.40', 456, 293, 196, 93, 507, 61, 185, 119,
       72, 66, 249, 180, 173, 109, 251, 29.99, 122, 129, 23.99, 34.99,
       3.99, 'From £40.00', 273, 445, 195, 139, 83, 159, 703, 146, 119.99,
       580, 102, 179, 299, 107, 209, 277, 77, 76, 47, 84.99, 88, 290, 236,
       230, 98, 156, 203, 275, 67, 44.99, 222, 189, 104, 400, 12.5, 304,
       142, 239, 194, 202, 92, 'From £29.00', 204, 31, 417, 555, 228, 171,
       24.99, 155, 39.99, 291, 208, 41, 420, 

1. Prices that are NAN as they are sold without providing prices
2. Price 14.3, seems unnecessary to sell for at neither half-price or 1p less than the nearest pound
3. Price 15.20 

ran out of time but would have checked average category prices and compared each price against it

#  Give 2 examples of products in the dresses subcategory where the price of the product is greater than 4 mul"plied by the upper quar"le price of the dresses subcategory

no time